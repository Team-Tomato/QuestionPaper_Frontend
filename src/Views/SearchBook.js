import React, { Component } from 'react';
import '../Styles/style.css'
import Loader from 'react-loading';
import { Form, Row, Col, Container, Card, CardBody, Table } from 'reactstrap'
import ReactPaginate from 'react-paginate';
import '../Styles/pagination.css'
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import styled from 'styled-components';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faSortAlphaDown } from '@fortawesome/free-solid-svg-icons'

const StyledTextField = styled(TextField)`{
  label.Mui-focused {
    color: violet;
  }
  .MuiOutlinedInput-root {
    &.Mui-focused fieldset {
      border-color: violet;
    }
  }`;

const valreg=RegExp(/^\s+$/)

class Project extends Component {
  constructor(){
    super()
  this.state = {
    query: '',
    loading: false,
    person: [],
    offset:0,
    perPage:10,
    currentPage:0,
    noData:false,
    sort:false,
    choice:'',
    prev:''
  };
  this.handlePageClick = this.handlePageClick.bind(this)
  this.handleSubClick=this.handleSubClick.bind(this)
}
  onChange = e => {
    const { value } = e.target;
    let error= this.state.error;  
    if (value.length === 0 || valreg.test(value))
    {
      error = true;
    }
    else
    {
      error = false;
    }        
    this.setState({error,value },()=>console.log(this.state)) ;    
    this.setState({
      query: value,
      person: []
    });
  };

  handlePageClick=(e)=>{
    const selectedPage = e.selected;
    const offset = selectedPage * this.state.perPage;

    this.setState({
        currentPage: selectedPage,
        offset: offset
    });
  }
 
  async search(query) {
    const url = "https://teamtomato.herokuapp.com/api/v1/book/search?search_str=" + query //fetch the specific book
    const response = await fetch(url)
    const data = await response.json();
    this.setState({
      person: data,
      loading: false,
      sort:false,
      choice:'',
      prev:''
    })
    if(this.state.person.length==0)
    {
      this.setState({noData:true})
    }else{this.setState({noData:false})}
  }

  handleLogin = e => {
    const query = this.state.query
    this.setState({ loading: true })
    this.search(query)
  }

  handleSubClick=(e,value)=>{
    e.preventDefault();
  this.setState({sort:true,
    prev:this.state.choice,
  choice:value})
  }
  render() {
    const {error} = this.state;
    let BookContainer
    let table = []
    if (this.state.loading === false) {
      if (this.state.person !== [] && (this.state.person).length !== 0) {
        if(this.state.sort)
        {
          if(this.state.prev !== this.state.choice)
          {
          const v=this.state.choice
          const sorted=this.state.person.sort(function(a,b){
            if(a[v]<b[v])
            return -1;
            if(a[v]<b[v])
            return 1;
            return 0;
        }); 
        const slice=sorted.slice(this.state.offset, this.state.offset + this.state.perPage) 
        table = slice.map((data, index) => {
          return (
            <tr key={index}>
            <td>{data.title}</td>
            <td style={{textAlign:"center"}}>{data.author}</td>
            <td>{data.isbn}</td>
            <td style={{textAlign:"center"}}>{data.publisher}</td>
            <a href={data.url} target="blank" className="violet"><td>{data.url}</td></a>
          </tr>
          )
        })}
        else{
          const v=this.state.choice
          const sorted=this.state.person.sort(function(a,b){
            if(a[v]>b[v])
            return -1;
            if(a[v]>b[v])
            return 1;
            return 0;
        }); 
        const slice=sorted.slice(this.state.offset, this.state.offset + this.state.perPage) 
        table = slice.map((data, index) => {
          return (
            <tr key={index}>
            <td>{data.title}</td>
            <td style={{textAlign:"center"}}>{data.author}</td>
            <td>{data.isbn}</td>
            <td style={{textAlign:"center"}}>{data.publisher}</td>
            <a href={data.url} target="blank" className="violet"><td>{data.url}</td></a>
          </tr>
          )
        })
        }
        }
        else{
        const slice=this.state.person.slice(this.state.offset, this.state.offset + this.state.perPage)
        table = slice.map((data, index) => {
          return (
            <tr key={index}>
              <td>{data.title}</td>
              <td style={{textAlign:"center"}}>{data.author}</td>
              <td>{data.isbn}</td>
              <td style={{textAlign:"center"}}>{data.publisher}</td>
              <a href={data.url} target="blank" className="violet"><td>{data.url}</td></a>
            </tr>
          )
        })}
        BookContainer =
          <Container>
            <Table striped hover responsive>
              <thead>
                <tr>
                  <th><Button onClick={(event) => { this.handleSubClick(event, 'title')}} style={{backgroundColor: "violet",color:"white",width:100}}>Title <FontAwesomeIcon icon = {faSortAlphaDown}/></Button></th>
                  <th><Button onClick={(event) => { this.handleSubClick(event, 'author')}} style={{backgroundColor: "violet",color:"white",width:100}}>Author <FontAwesomeIcon icon = {faSortAlphaDown}/></Button> </th>
                  <th><Button onClick={(event) => { this.handleSubClick(event, 'isbn')}} style={{backgroundColor: "violet",color:"white",width:120}}>ISBN <FontAwesomeIcon icon = {faSortAlphaDown}/></Button></th>
                  <th><Button onClick={(event) => { this.handleSubClick(event, 'publisher')}} style={{backgroundColor: "violet",color:"white",width:120}}>Publisher <FontAwesomeIcon icon = {faSortAlphaDown}/></Button></th>
                  <th>URL</th>
                </tr>
              </thead>
              <tbody>
                {table}
              </tbody>
            </Table>
          </Container>
      }
      else {
        BookContainer = <h6></h6>
      }
    }
    else {
      BookContainer = <div style={{
        position: 'absolute', left: '50%', top: '77%',
        transform: 'translate(-50%, -50%)'
      }}>
        <Loader type={"bars"} color={"black"} />
      </div>
    }
    const count=Math.ceil(this.state.person.length / this.state.perPage);
    return (
      <div>
        <Card className="gradient">
          <CardBody className="welcome-title" style={{
            position: 'absolute', left: '50%', top: '50%',
            transform: 'translate(-50%, -50%)',
            color: 'white'
          }}>
            <h5>Books destination for Integrated students</h5>
          </CardBody>
        </Card>
        <br />
        <Container>
          <Card className="correctMargin">
            <CardBody className="removeIndent">
              <Form>
                <Row>
                  <Col sm={6} md={8} lg={10} className="addIndent">
                  <StyledTextField fullWidth type="text" label="Enter the Title or Author name"  onChange={this.onChange} id="outlined-size-small" variant="outlined"size="small"/>
                    {error === true && (
                      <div className="errormessage">title or author name is required</div>
                    )}
                  </Col>
                  <Col sm={6} md={4} lg={2} className="addIndent">
                    <Button type="submit" className="col-md-12" disabled={!this.state.value || this.state.value.trim().length === 0} style={{ backgroundColor: "violet",color:"white"}} variant="contained" block onClick={this.handleLogin}>Search</Button>
                  </Col>
                </Row>
              </Form>
            </CardBody>
          </Card>
          <div>
          </div>
        </Container>
        {BookContainer}
      
        {this.state.noData ? (
          <div><br/>
            <h5 className="centerIt">Sorry,we are not able to find the book you are looking for</h5>
            </div>
            
            ):(count>1 ? (
              <div>
                 <ReactPaginate
                    previousLabel={"prev"}
                    nextLabel={"next"}
                    breakLabel={"..."}
                    breakClassName={"break-me"}
                    pageCount={count}
                    marginPagesDisplayed={2}
                    pageRangeDisplayed={5}
                    onPageChange={this.handlePageClick}
                    containerClassName={"pagination"}
                    subContainerClassName={"pages pagination"}
                    activeClassName={"active"}/>
              </div>
            ):(<div></div>)
            )}
      </div>
    )
  }
}

export default Project;