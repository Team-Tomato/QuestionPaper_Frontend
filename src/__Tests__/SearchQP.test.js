import 'jsdom-global/register';
import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import fetchMock from 'jest-fetch-mock';

import SearchQP from '../Views/searchQP';
import SearchBook from '../Views/SearchBook';
import { mount, shallow } from 'enzyme';
import { create } from 'react-test-renderer';

Enzyme.configure({ adapter: new Adapter() })
fetchMock.enableMocks();

describe('Testing Search Qp page',()=>{
    test('Matches snapshots',()=>{
        const snap=create(<SearchQP />);
        expect(snap.toJSON()).toMatchSnapshot();
    });
    test.skip('testing input',()=>{
        const QP = shallow(<SearchQP  />)
        const container = QP.find('Container');
        expect(container).toHaveLength(2);
        let card = container.find('Card');
        expect(card).toHaveLength(1);
        let card_body = card.find('CardBody'); 
        expect(card_body).toHaveLength(1);
        let form = card_body.find('Form').find('Row');
        expect(form).toHaveLength(1);
        let col = form.find('Col');
        expect(col).toHaveLength(2);
        const inp = col.at(0).find('StyledTextField');
        expect(inp).toHaveLength(1);
        inp.simulate('change',{
            target:{
                value: 'testing'
            }
        })
        expect(inp).toBe('testing');
        //.getDOMNode().value = 'testing';
        //QP.find('StyledTextField').simulate('change');
        //expect(inp).toEqual('testing');        
    });
    test('testing year button',()=>{
        const qp = mount(<SearchQP></SearchQP>);
        const container = qp.find('Container');
        expect(container).toHaveLength(2);
        const row =container.find('Row');
        expect(row).toHaveLength(2);
        const col = row.find('Col');
        expect(col).toHaveLength(6);
        const a = col.find('a');
        expect(a).toHaveLength(4);
        let ref = a.at(0).prop('href');
        let name = a.at(0).find("Button");
        //expect(name).toHaveLength(1);
        //expect(name.text()).toEqual('2018-2023');
        expect(ref).toEqual("https://github.com/Team-Tomato/Learn/blob/master/QP%20Data/2018batch.md");
        //expect(document.querySelector("a").getAttribute("href").toBe("https://github.com/Team-Tomato/Learn/blob/master/QP%20Data/2017batch.md"));
    }); 
    test('Calling Onsubmit method',()=>{
        const qp = mount(<SearchQP />);
        const preventDefault = jest.fn();
        const query = 'Testing';
        qp.setState({
            query
        });
        qp.find('Form').simulate('submit',{preventDefault});
        expect(preventDefault).toBeCalled();
    });   
});


describe("Fetching data using jest-fetch-mock",()=>{
    const handlesubmit = jest.fn('submit','Testing');
    test('on successful fetch request',async ()=>{
        fetch.mockResponse(()=>handlesubmit().then(res=> 'ok'));

        //const query = 'Testing';
    //    const data = await handlesubmit('submit', 'Testing');
    //    expect(handlesubmit).toBeCalledTimes(1);
        expect(fetch).toHaveBeenCalledTimes(1);
        expect(fetch).toHaveBeenCalledWith("https://teamtomato.herokuapp.com/api/v1/question/search?search_str=", {
            "method" : 'GET',
            "headers" : {
              'Access-Control-Allow-Origin': '*',
              'Content-Type': 'application/json',
              'Access-Control-Allow-Methods': 'GET,PUT,POST,DELETE'
            }
        });
    })
    test('on a failed fetch request', async()=>{
        fetch.mockReject(()=>handlesubmit().then(res => Promise.reject(new Error('Network response was not ok.'))))
        expect(fetch).toHaveBeenCalledTimes(1);
        expect(fetch).toHaveBeenCalledWith("https://teamtomato.herokuapp.com/api/v1/question/search?search_str=", {
            "method" : 'GET',
            "headers" : {
              'Access-Control-Allow-Origin': '*',
              'Content-Type': 'application/json',
              'Access-Control-Allow-Methods': 'GET,PUT,POST,DELETE'
            }
        })  
    })
})