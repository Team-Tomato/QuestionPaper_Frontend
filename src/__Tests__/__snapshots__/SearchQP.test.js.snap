exports[`Testing Search Qp page Matches snapshots 1`] = `
<div>
  <div
    className="gradient card">
    <div
      className="welcome-title card-body"
      style={
        Object {
          "color": "white",
          "left": "50%",
          "position": "absolute",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        }
      }>
      <h5>
        Search for integrated M.Sc question papers
      </h5>
    </div>
  </div>
  <div
    className="container">
    <br />
    <div
      className="correctMargin card">
      <div
        className="removeIndent card-body">
        <form
          className=""
          onSubmit={[Function]}>
          <div
            className="addSpace row">
            <div
              className="addIndent col-sm-6 col-md-8 col-lg-10">
              <div
                className="MuiFormControl-root MuiTextField-root sc-AxjAm kClKlr MuiFormControl-fullWidth">
                <label
                  className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
                  data-shrink={false}
                  htmlFor="outlined-size-small keywordSearch"
                  id="outlined-size-small keywordSearch-label">
                  Search subject name or staff
                </label>
                <div
                  className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
                  onClick={[Function]}>
                  <input
                    aria-describedby={undefined}
                    aria-invalid={false}
                    autoComplete={undefined}
                    autoFocus={false}
                    className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
                    defaultValue={undefined}
                    disabled={false}
                    id="outlined-size-small keywordSearch"
                    name="keywordSearch"
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyDown={undefined}
                    onKeyUp={undefined}
                    placeholder={undefined}
                    readOnly={undefined}
                    required={false}
                    rows={undefined}
                    type="text"
                    value={undefined} />
                  <fieldset
                    aria-hidden={true}
                    className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
                    style={undefined}>
                    <legend
                      className="PrivateNotchedOutline-legendLabelled-3">
                      <span>
                        Search subject name or staff
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              className="addIndent col-sm-6 col-md-4 col-lg-2">
              <button
                block={true}
                className="MuiButtonBase-root MuiButton-root MuiButton-contained col-md-12 Mui-disabled Mui-disabled"
                disabled={true}
                onBlur={[Function]}
                onClick={undefined}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                style={
                  Object {
                    "backgroundColor": "violet",
                    "color": "white",
                  }
                }
                tabIndex={-1}
                type="submit">
                <span
                  className="MuiButton-label">
                  Search
                </span>
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
  <br />
  <div
    className="container">
    <h5
      className="centerIt">
      Yearwise question paper collection status
    </h5>
    <div
      className="row">
      <div
        className="col-sm-6 col-md-3 col-lg-3">
        <a
          href="https://github.com/Team-Tomato/Learn/blob/master/QP%20Data/2018batch.md"
          style={
            Object {
              "textDecoration": "none",
            }
          }
          target="_blank">
          <button
            block={true}
            className="MuiButtonBase-root MuiButton-root MuiButton-contained col-md-12"
            disabled={false}
            onBlur={[Function]}
            onClick={undefined}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "backgroundColor": "violet",
                "color": "white",
                "textDecoration": "none",
              }
            }
            tabIndex={0}
            type="button">
            <span
              className="MuiButton-label">
              2018-2023
            </span>
          </button>
        </a>
      </div>
      <div
        className="col-sm-6 col-md-3 col-lg-3">
        <a
          href="https://github.com/Team-Tomato/Learn/blob/master/QP%20Data/2017batch.md"
          style={
            Object {
              "textDecoration": "none",
            }
          }
          target="_blank">
          <button
            block={true}
            className="MuiButtonBase-root MuiButton-root MuiButton-contained col-md-12"
            disabled={false}
            onBlur={[Function]}
            onClick={undefined}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "backgroundColor": "violet",
                "color": "white",
                "textDecoration": "none",
              }
            }
            tabIndex={0}
            type="button">
            <span
              className="MuiButton-label">
              2017-2022
            </span>
          </button>
        </a>
      </div>
      <div
        className="col-sm-6 col-md-3 col-lg-3">
        <a
          href="https://github.com/Team-Tomato/Learn/blob/master/QP%20Data/2016batch.md"
          style={
            Object {
              "textDecoration": "none",
            }
          }
          target="_blank">
          <button
            block={true}
            className="MuiButtonBase-root MuiButton-root MuiButton-contained col-md-12"
            disabled={false}
            onBlur={[Function]}
            onClick={undefined}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "backgroundColor": "violet",
                "color": "white",
                "textDecoration": "none",
              }
            }
            tabIndex={0}
            type="button">
            <span
              className="MuiButton-label">
              2016-2021
            </span>
          </button>
        </a>
      </div>
      <div
        className="col-sm-6 col-md-3 col-lg-3">
        <a
          href="https://github.com/Team-Tomato/Learn/blob/master/QP%20Data/2015batch.md"
          style={
            Object {
              "textDecoration": "none",
            }
          }
          target="_blank">
          <button
            block={true}
            className="MuiButtonBase-root MuiButton-root MuiButton-contained col-md-12"
            disabled={false}
            onBlur={[Function]}
            onClick={undefined}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "backgroundColor": "violet",
                "color": "white",
                "textDecoration": "none",
              }
            }
            tabIndex={0}
            type="button">
            <span
              className="MuiButton-label">
              2015-2020
            </span>
          </button>
        </a>
      </div>
    </div>
  </div>
  <br />
  <h6 />
  <div />
</div>
`;
